[GTK][WPE] Use damaging information in the composition

Scony opened the pull request:
#### 84a1b08a7a8fe0f12baa697dacdc369feacc43bf
<pre>
[GTK][WPE] Use damaging information in the composition
<a href="https://bugs.webkit.org/show_bug.cgi?id=277337">https://bugs.webkit.org/show_bug.cgi?id=277337</a>

Reviewed by Carlos Garcia Campos.

This PR adds a mechanism that leverages the damage information to reduce the
amount of painting during composition.

The idea behind above mechanism is very simple - the individual damage information
is being tracked for every RenderTarget. Once the painting is being performed to the
certain RenderTarget, its individual damage information is used to clip painting to
modified rects thus reducing the amount of actual painting.

This PR also separates damage collection from painting to make the above possible.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/graphics/texmap/TextureMapperFPSCounter.h:
(WebCore::TextureMapperFPSCounter::isActive const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamage):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::collectDamageRecursive):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::addDamage):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::addSurfaceDamage):
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetSHMImage::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::addDamage):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:

Canonical link: <a href="https://commits.webkit.org/286854@main">https://commits.webkit.org/286854@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/18750e2e6c3d417f2b31259ba6b73ba8169686fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77118 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56153 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81694 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28400 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79235 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65301 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4449 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60437 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18496 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80185 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50391 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66200 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40744 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23698 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26723 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70303 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68908 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83103 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76396 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4498 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3030 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68808 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4654 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66173 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68065 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17003 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11950 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10037 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98649 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4444 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7260 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21585 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4464 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7899 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6223 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.h:
```diff
@@ -51,12 +51,14 @@ class CoordinatedGraphicsSceneClient {
 public:
     virtual ~CoordinatedGraphicsSceneClient() { }
     virtual void updateViewport() = 0;
+    virtual void damageRenderTargets(const WebCore::Damage&);
```
```suggestion
    virtual void damageRenderTargets(const WebCore::Damage&) = 0;
```

Scony commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
```diff
@@ -69,6 +71,9 @@ class ThreadedCompositor : public CoordinatedGraphicsSceneClient, public ThreadS
         virtual void didRenderFrame(uint32_t, const WebCore::Damage&) = 0;
         virtual void displayDidRefresh(WebCore::PlatformDisplayID) = 0;
+
+        virtual void damageRenderTargets(const WebCore::Damage&);
```
```suggestion
        virtual void damageRenderTargets(const WebCore::Damage&) = 0;
```

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -1045,6 +1070,22 @@ void TextureMapperLayer::paintFlattened(TextureMapperPaintOptions& options)
+    collectDamageSelf(options);
+
+    if (m_children.isEmpty())
+        return;
```
This is not needed, the for loop below will do nothing anyway

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -480,6 +465,46 @@ void TextureMapperLayer::paintPreserves3DHolePunch(TextureMapperPlatformLayer* c
+        // calls from CoordinatedGraphicsLayer.
+        if (m_propagateDamage == Damage::ShouldPropagate::Yes) {
+            if (m_damage.isInvalid())
+                recordDamage(layerRect(), transform, options);
```
```suggestion
                recordDamage(targetRect, transform, options);
```

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -480,6 +465,46 @@ void TextureMapperLayer::paintPreserves3DHolePunch(TextureMapperPlatformLayer* c
+
+    if (m_propagateDamage == Damage::ShouldPropagate::Yes) {
+        // Layers with content layer are always fully damaged for now...
+        recordDamage(layerRect(), transform, options);
```
```suggestion
        recordDamage(targetRect, transform, options);
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -67,13 +69,48 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
+        m_fpsCounter.updateFPSAndDisplay(*m_textureMapper, clipRect.location(), matrix);
+        m_textureMapper->endClip();
+        m_textureMapper->endPainting();
+    }
```
Why do we need to separate the fps counter painting here now?

Scony replied:
FPS counter is not taking part in damaging the frame, so if damage propagation is enabled, now it's possible that the clipping region will be limited to the area that does not include FPS counter and therefore counter won't be actually painted (it will be painted outside the clipping region).

Technically - as the FPS counter is not working with skia atm. - this change currently prevents regression only in cairo builds.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -30,12 +30,14 @@
-CoordinatedGraphicsScene::CoordinatedGraphicsScene(CoordinatedGraphicsSceneClient* client, Damage::ShouldPropagate propagateDamage)
+CoordinatedGraphicsScene::CoordinatedGraphicsScene(CoordinatedGraphicsSceneClient* client, const WebCore::Settings& settings, Damage::ShouldPropagate propagateDamage)
     : m_client(client)
+    , m_settings(settings)
```
I don't think we should keep settings as a member. Let's just pass the settings we need to the methods using them, or to the constructor and keep the setting value, not the settings object, liked currently do for m_propagateDamage

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -67,13 +69,48 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
 
+    WTFBeginSignpost(this, CollectDamage);
+    currentRootLayer->collectDamage(*m_textureMapper);
+    WTFEndSignpost(this, CollectDamage);
```
Should we collect damage only when enabled in the settings?

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -67,13 +69,48 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
+        // FIXME: Decide what to do with valid-yet-empty damage (bounds:0,0,0,0).
+        m_client->damageRenderTargets(frameDamage);
+
+        const Damage& damageSinceLastRenderTargetUse = m_client->renderTargetDamage();
```
I don't understand this, the names damageRenderTarget and renderTargetDamage are a bit confusing, and they are called one after the other, couldn't it be just one?

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -67,13 +69,48 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
+
+        const Damage& damageSinceLastRenderTargetUse = m_client->renderTargetDamage();
+        if (!damageSinceLastRenderTargetUse.isInvalid())
+            actualClipRect = (FloatRoundedRect)damageSinceLastRenderTargetUse.bounds();
```
Don't do C style casts

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
```diff
@@ -61,6 +62,9 @@ class AcceleratedSurface {
     virtual void didRenderFrame(WebCore::Region&&) { }
 
+    virtual void damageRenderTargets(const WebCore::Damage&) { };
+    virtual const WebCore::Damage& renderTargetDamage() { ASSERT_NOT_REACHED(); return WebCore::Damage::invalid(); };
```
RenderTarget is an implementation detail specific to the DMABuf surface, I think these should not use it. What about just `addDamage()` and `damage()` ?

carlosgcampos replied:
Also, I guess the old api will hit this assert if the setting is enabled, I think it should be pure virtual and implemented in old api, or remove the assert.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -657,6 +671,14 @@ void AcceleratedSurfaceDMABuf::didRenderFrame(WebCore::Region&& damage)
 
     m_target->didRenderFrame();
     WebProcess::singleton().parentProcessConnection()->send(Messages::AcceleratedBackingStoreDMABuf::Frame(m_target->id(), WTFMove(damage), WTFMove(renderingFence)), m_id);
+    if (m_target)
```
This can't be nullptr here, it's checked before and used without checks in the previous lines.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -657,6 +671,14 @@ void AcceleratedSurfaceDMABuf::didRenderFrame(WebCore::Region&& damage)
     m_target->didRenderFrame();
     WebProcess::singleton().parentProcessConnection()->send(Messages::AcceleratedBackingStoreDMABuf::Frame(m_target->id(), WTFMove(damage), WTFMove(renderingFence)), m_id);
+    if (m_target)
+        m_target->resetDamage();
```
Maybe this could be done automatically in Target::didRenderFrame? and we don't need resetDamage()?

Scony committed:
[GTK][WPE] Use damaging information in the composition
https://bugs.webkit.org/show_bug.cgi?id=277337

Reviewed by Carlos Garcia Campos.

This PR adds a mechanism that leverages the damage information to reduce the
amount of painting during composition.

The idea behind above mechanism is very simple - the individual damage information
is being tracked for every RenderTarget. Once the painting is being performed to the
certain RenderTarget, its individual damage information is used to clip painting to
modified rects thus reducing the amount of actual painting.

This PR also separates damage collection from painting to make the above possible.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/graphics/texmap/TextureMapperFPSCounter.h:
(WebCore::TextureMapperFPSCounter::isActive const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamage):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::collectDamageRecursive):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::addDamage):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::addSurfaceDamage):
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::RenderTargetSHMImage::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::addDamage):
(WebKit::AcceleratedSurfaceDMABuf::addDamage):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:

Canonical link: https://commits.webkit.org/286854@main
